compiler/src/typed/cmi_format.re

```diff
@@ -142,22 +138,32 @@ module CmiBinarySection =
     let accepts_version = ({major}) => major == 1;
   });
 
-let read_cmi = filename => {
-  let ic = open_in_bin(filename);
-  switch (CmiBinarySection.load(ic)) {
-  | Some(cmi) =>
-    close_in(ic);
-    cmi;
-  | None
-  | exception End_of_file
-  | exception (Invalid_argument(_))
-  | exception (Failure(_)) =>
-    close_in(ic);
+let read_cmi = (ic, filename): cmi_infos => {
+  let read_magic = really_input_string(ic, 4);
+  if (read_magic != magic) {
     raise(Error(Corrupted_interface(filename)));
-  | exception (Error(e)) =>
-    close_in(ic);
-    raise(Error(e));
   };
+  let version_length = input_binary_int(ic);
```

Comment:
Would it make sense to move some of this logic into its own module so we can re-use the checks between this file and emit_mod? This would help ensure our checks are staying the same and our object file parsing is the same.

Reply:
I wanted to do that, but there's some annoying cyclic dependency issues (Grain_codegen depends on Grain_typed, and this would introduce a dependency on Grain_codegen). I admit that this is one annoying place where the two need to be kept in sync, but we'll very very rarely change it and it'll be obviously broken if we do.

Reply:
Do we maybe want to open an issue for this?

Reply:
Not particularly; I meant that in a "I don't really plan to do anything about this because I don't think it's a big deal" way. There's no good reason to refactor the whole compiler just for this. The only thing we could do that would be slightly better is implementing the half that doesn't rely on the mashtree somewhere and importing that into emitmod and implementing the other half, but I don't think that's really all that much better.